docs/proposals/event-notification.md

```diff
@@ -0,0 +1,143 @@
+
+Note that unlike other machine-readable outputs used in odo, each individual line is a fully complete and parseable JSON document, allowing events to be streamed and processed by the consuming tool one-at-a-time, rather than waiting for all the events to be received before being parseable (which would be required if the entire console output was one single JSON document, as is the case for other odo machine-readable outputs.)
+
+### Detailed build status via JSON+custom markup
+
+For detailed build status, it is proposed that devfile writers may *optionally* include custom markup in their devfile actions which indicate a detailed build status:
+- If a dev file writer wanted to communicate that the current command/action were compiling the application, they would insert a specific markup string (`#devfile-status#`) at the beginning of a console-outputted line, and then between those two fields would be a JSON object with a single field `buildStatus`:
+  - For example: `#devfile-status# {"buildStatus":"Compiling application"}` would then communicate that the detailed build status should be set to `Compiling application`.
+- Since this line would be output as container stdout, it would be included as a `logText` JSON event, and the consuming tool can look for this markup string and parse the simple JSON to extract the detailed build status.
+- Feedback welcome around exact markup text format.
+
+The build step (running as a bash script, for example, invoked via an action) of a devfile might then look like this:
+```
+#!/bin/bash
+(...)
+echo "#devfile-status# {'buildStatus':'Compiling application'}
+mvn compile
+echo "#devfile-status# {'buildStatus':'Running unit tests'}
+mvn test
+```
+
+This 'detailed build status' markup text is *entirely optional*: if this markup is not present, the odo tool can still determine build succeeded/failed and build running/not-running using the other `odo push -o json` JSON events. 
+
+### App status notification via `odo component status -o json` and `odo log --follow`
+
+In general, within the execution context that odo operates, there are a few ways for us to determine the application status:
+1) Can the application be pinged at its exposed URL? 
+2) What state is the container in? (running/container creating/restarting/etc -- different statuses between local and Kube but same general idea)
+3) In the application log, specific hardcoded text strings can be searched for (for example, OpenLiberty outputs defined status codes to its log to indicate that an app started.)
+
+Ideally, we would like for odo to provide consuming tools with all 3 sets of data. Thus, as proposed:
+- 1 and 2 are handled by a new `odo component status -o json` command, described here.
+- 3 is handled by the existing unmodified `odo log --follow` command.
+
+The new proposed `odo component status -o json` command will:
+- Be a *long-running* command that will continue outputing status until it is aborted by the parent process.
```

Comment:
Good point, and actually, in an earlier version of this design proposal this was called `odo component status -o json --follow`, but then I simplified it to remove `--follow`.

Would you prefer something like `odo component status -o json --follow`? This definitely makes the long-running behaviour of the command more clear.

The other item it sounds like you may be proposing in your above comment is combining the `odo log` and `odo component status` commands? I had considered this, but ultimately realized there is a case where the lifecycle of the two commands should differ: when the user adds a new URL to an existing component, the `odo component status` process needs to be restarted (eg to pick up the new URL), but the `odo log --follow` process does not need to be restarted (because the container/pod remain the same).

Reply:
> Would you prefer something like `odo component status -o json --follow`? This definitely makes the long-running behaviour of the command more clear.

I agree that this makes the long-running behaviour more obvious to the user. However, I was thinking of `odo log --follow -o json` but your later point about `odo log --follow` being something that shouldn't need a restart makes sense as well and. So, I guess, `odo component status -o json --follow` would be more appropriate.